trinity/sync/full/chain.py

```diff
@@ -442,29 +380,444 @@ class FastChainSyncer(BaseHeaderChainSyncer):
         peer.sub_proto.send_block_headers(headers)
 
 
-class RegularChainSyncer(FastChainSyncer):
+class BlockPersistPrereqs(enum.Enum):
+    StoreBlockBodies = enum.auto()
+    StoreReceipts = enum.auto()
+
+
+class FastChainSyncer(BaseBodyChainSyncer):
+    """
+    Sync with the Ethereum network by fetching block headers/bodies and storing them in our DB.
+
+    Here, the run() method returns as soon as we complete a sync with the peer that announced the
+    highest TD, at which point we must run the StateDownloader to fetch the state for our chain
+    head.
+    """
+    db: AsyncChainDB
+
+    def __init__(self,
+                 chain: AsyncChain,
+                 db: AsyncHeaderDB,
+                 peer_pool: PeerPool,
+                 token: CancelToken = None) -> None:
+        super().__init__(chain, db, peer_pool, token)
+
+        # queue up any idle peers, in order of how fast they return receipts
+        self._receipt_peers = WaitingPeers(commands.Receipts)
+
+        # Track receipt download tasks
+        # - arbitrarily allow up to 4 requests-worth of headers queued up
+        # - try to get receipts from lower block numbers first
+        self._receipt_tasks = TaskQueue(MAX_RECEIPTS_FETCH * 4, attrgetter('block_number'))
+
+        # track when both bodies and receipts are collected, so that blocks can be persisted
+        self._block_persist_tracker = OrderedTaskPreparation(
+            BlockPersistPrereqs,
+            id_extractor=attrgetter('hash'),
+            # make sure that a block is not persisted until the parent block is persisted
+            dependency_extractor=attrgetter('parent_hash'),
+        )
+
+    async def _run(self) -> None:
+        head = await self.wait(self.db.coro_get_canonical_head())
+        self._block_persist_tracker.set_finished_dependency(head)
+        self.run_task(self._launch_prerequisite_tasks())
+        self.run_task(self._assign_receipt_download_to_peers())
+        self.run_task(self._assign_body_download_to_peers())
+        self.run_task(self._persist_ready_blocks())
```

Comment:
So in theory, most of these background tasks are required for the service to be operational/functional.  Using `run_task` allows for them to exit at which point the client will keep running but sync will stop working.  Thoughts on how we can address this?  Something like `BaseService.run_daemon_task()`?